helix-core/src/graphemes.rs

```diff
@@ -341,3 +425,101 @@ impl<'a> Iterator for RopeGraphemes<'a> {
         }
     }
 }
+
+/// A highlight compressed Cow<'a, str> that holds
+/// atmost u31::MAX bytes and is readonly
+pub struct GraphemeStr<'a> {
+    ptr: NonNull<u8>,
+    len: u32,
+    phantom: PhantomData<&'a str>,
+}
```

Comment:
Could this just be 

```rust
enum GraphemeStr {
    Owned(Box<str>),
    Borrowed(&'a str),
}
```

I think `Box<>` has the niche optimization equivalent to `NonNull` so it should consume the same amount of space. `len` then wouldn't be computed but would be one deref away

Reply:
I guess not: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=cd620bdefa74b438a8d92416e5c000fa

Reply:
The enum is functionally equivalent to `GraphemeStr` but not in size because niche optimization only work for empty enum variants. So the following would be two words long:

``` rust
enum GraphemeStr {
    Foo(Box<str>),
    Bar
}
```

The reason for this is that you only have a single bit pattern (`0`) here that is not already occupied. So the compiler performs roughly the following transformation for a match:

``` rust
match grapheme_str{
   GraphemeStr::Foo(foo) => println!("{foo}"),
   GraphemeStr::Bar => println!("bar"),
}
```

``` rust

  if transmute(grapheme_str) == [0usize; 2]{
     println!("bar")
  }else{
     println!("{}", transmute::<GraphemeStr, Box<str>>(grapheme_str))
  }
}
```

This doesn't work when both variants are occupied tough.

The reason the `GraphemeStr` is  is that I reduced that maximum permissible length of a grapheme to `2^31-1` (31 bits).
Theoretically a grapheme could be infinitly many bytes long but I think limiting the size to 2GB here is very reasonable.

This reduction in maximum length leaves us with enough space to store the discriminant.